core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/marshall/HeaderMarshaller.java

```diff
@@ -79,4 +72,36 @@ public void marshall(T val, JsonMarshallerContext context, String paramName, Sdk
             context.request().appendHeader(paramName, converter.convert(val, sdkField));
         }
     }
+
+    private static class SimpleListHeaderMarshaller implements JsonMarshaller<List<?>> {
+        private Map<Class<?>, Predicate> skipPredicateMap = new HashMap<>();
+
+        {
+            skipPredicateMap.put(String.class,
+                                 k -> k instanceof String && StringUtils.isBlank((String) k));
+        }
+
+        @Override
+        public void marshall(List<?> list, JsonMarshallerContext context, String paramName, SdkField<List<?>> sdkField) {
+            // Null or empty lists cannot be meaningfully (or safely) represented in an HTTP header message since header-fields
+            // must typically have a non-empty field-value. https://datatracker.ietf.org/doc/html/rfc7230#section-3.2
+            if (isNullOrEmpty(list)) {
+                return;
+            }
+            SdkField memberFieldInfo = sdkField.getRequiredTrait(ListTrait.class).memberFieldInfo();
+            for (Object listValue : list) {
+                if (shouldSkipElement(listValue)) {
+                    continue;
+                }
+                JsonMarshaller marshaller = context.marshallerRegistry().getMarshaller(MarshallLocation.HEADER, listValue);
+                marshaller.marshall(listValue, context, paramName, memberFieldInfo);
+
+            }
+        }
+
+        private boolean shouldSkipElement(Object listValue) {
+            // Null values are not skipped since they have predefined Null marshallers.
```

Comment:
Not sure I understand this comment. This fix *does* skip null values right?

Reply:
I was speaking about this https://github.com/aws/aws-sdk-java-v2/blob/55f314fbe4af9cfac0c67cbff505476a5c344a5f/core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/marshall/JsonProtocolMarshaller.java#L123